###Titanic Python

1/May/2019

*"first catch your hare"*

In our terms get the data and understand the objective. 


In [14]:
# simple statment to confirm that the notebook processing is active
# for Google colab - it often takes a few seconds to get things started 
# this simple step with its clear output provides a visual check
print ("Titanic - Python")

Titanic - Python


### Imports

One of the strengths of Python is the range of external libraries which can be used.

An important mathematical library - which implements multidimensional arrays and many mathods for manipulating them is numpy - which by convention we import as 'np'  

For data analysis there is an important library called pandas which implements a dataframe which can be used to store tabular data and implements many useful methods for manipulating these dataframes - by convention we import it as 'pd'



In [0]:
import pandas as pd
import numpy as np

Import data from Google drive : 

For the next section of code to work you have to have copied the relevant files to your
Google Drive.

And as far as I know it only works from Google Colab - which is where I created this notebook.

I have commented the code to prevent it running and generating an error if the Google Drive link is 
not in place -  it requires a Google Drive authorisation every session.

Clearly if you uncomment and use this code then you need to skip or comment out the next section.


In [0]:
# from google.colab import drive
# drive.mount('/content/gdrive')

# !ls /content/gdrive
# !ls -l /content/gdrive/My\ Drive
# !ls -l /content/gdrive/My\ Drive/Colab\ Notebooks
# !ls -l /content/gdrive/My\ Drive/Colab\ Notebooks/Titanic

# titanic_known = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/Titanic/titanic_known.csv')
# print (titanic_known.describe())
# titanic_unknown = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/Titanic/titanic_unknown.csv')
# print (titanic_unknown.describe())

Import data from a web source

The below is a simple load directly from a web source - which is fine for small datasets but not recommended if the amount of data gets large.
Any webserver will work.

I use github to store my working code - you can access files in a github repository directly using the https://raw.githubusercontent.com/ *username*/...... URL. 

Github also provides a 'special' repository name of *username*.github.io to allow you to create a repository which works as a static webserver (i.e. no server-side processing) - and this is perfectly serviceable for loading data into an object using the pandas read_csv method.

Pandas dataframes have a read_cvs method for reading csv sources - there are other simlar emthods for reading similar sources all of which can take a web URL as the location as well as a local file.

After reading file into a pandas dateframe we use len to get the number of rows and the we use the datframe method count to get the number of entries in each collum - and the collum names - this is pretty much a  *'have I got the data I expect?'*  verification.  And it gives us some basic insight into our data.

Whereas the whole purpose of using a machine to analyse data is to avoid much of the associated drudge work - it is hugely beneficial to have a good understanding of the data itself - particularly important is to make sure that the meaning of data is precise - specifically in this case we will have cause to ask what exactly does the Fare column contain.  And the asnwer to that question actually gives us an additional piece of data that we don't know that we have at the moment and that improves the performance of some of our models significantly.

To this end we will spend a bit of time just looking at the data in various ways.

In [17]:
titanic_known = pd.read_csv("https://raw.githubusercontent.com/johnsl01/Titanic_Python/master/titanic_known.csv", sep=",")

# titanic_known = pd.read_csv("https://johnsl01.github.io/titanic/titanic_known.csv", sep=",")
print ("Known :")
print (str(len(titanic_known)) + " Rows")
print (titanic_known.count())

titanic_unknown = pd.read_csv("https://johnsl01.github.io/titanic/titanic_unknown.csv", sep=",")
print ("\nUnknown :") 
print (str(len(titanic_unknown)) + " Rows")
print (titanic_unknown.count())

Known :
891 Rows
PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

Unknown :
418 Rows
PassengerId    418
Pclass         418
Name           418
Sex            418
Age            332
SibSp          418
Parch          418
Ticket         418
Fare           417
Cabin           91
Embarked       418
dtype: int64


#### head and tail

Our data is a bit unreal at the moment - we know the column names and the count of data in each - and we know that some ages and a lot of cabins are missing.

And we know that the 'known' dataframe has an extra column; 'Survived' which we expect to tell us what the fate of that passenger was - which is obviously missing from the 'unknown' dataframe.

One of the simplest things to do with data is to inspect it row by row and the starting point in this is usually to use the head and tail methods.  Head and tail derive from two unix commands which appeared very early in unix - and were used to show the first and last lines in a file.  By default they returned 5 lines and this default has persisted for almost 50 years as they are reincarnated in different systems.

So pandas dataframes have a head and a tail method which do the same for dataframe rows : 

In [18]:
print ("Known : -  first 5 and last 2:")
print (titanic_known.head())
print (titanic_known.tail(2))

print ("UnkKnown : -  first 5 and last 2:")
print (titanic_unknown.head())
print (titanic_unknown.tail(2))

Known : -  first 5 and last 2:
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            3734

OK - so this tells us someting and also raises some concerns.

The encoding of the survived column is 1 and 0  (1 for survived and 0 for died) - this is the most common numeric for a boolean value - 'Y' and 'N' and 'T' and 'F' being the most common character encodings - it is important to be sure that we get the mapping correct.  

The Pclass column is the passenger class - the ship was divided into three different classes - 1st, 2nd and 3rd.  1st class passengers had luxurious cabins, they had open access to the upper decks and to the luxurious 'state' rooms of the ship - they were generally wealthier and as well as some having brought personal staff with them they had a lot of the ship's crew dedicated to their care.  We should expect that this will be a factor in survival - access and crew deference being significant factors. 2nd class had individual cabins but less deck access and restricted access to some areas.  3rd class passengers generally were below decks - had dormitory type accomodation and were packed in pretty tight - they did not have much deck access and were unable to get into much of the 1st and 2nd class areas.

The name is the passenger's name and includes a title Mr. Mrs. etc as well as maiden names for married women.  This is text data so while we will want to make use of it we have to be careful in how we processes it - we will definitely want to pull out the title as that will help us in some cases with filling in the missing ages - and we will make some use of the name information to try and derive some information to help us with understanding what size of a party each passenger was likely in.

Sex is text and is either 'male' or 'female'.  Any knowledge of the history of the titanic will likely include the phrase 'women and children first'.  We should expect that it is probably a factor in survival - and we can validate it pretty quickly.

Age is self explanatory - it has a lot of missing values - we will look at mechanisms for filling in these values from the rest of the data, we expect that age will be a factor in survival - but need to explore how it affects it - expectation is probably that the elderly may have been less likely to survive - but despite being less able we expect the young may have survived both from having an adult caring for them and from the 'women and children first' policy - which is as yet unvalidated.

Lets look at our spread of values in the numeric columns using the describe method.

The PassengerID is just an incrementing counter which runs continuously from the knowns 1 through 891 and the unknowns (892 through 1309) - so the numerical analysis is pretty meaningless.

The counts are the same as above, the missing values in the ages column are clear.

In [19]:
print ("Known :")
print (titanic_known.describe())
print ("\nUnknown :") 
print (titanic_unknown.describe())

Known :
       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  

Unknown :
       PassengerId      Pclass         Age       SibSp       P

We have two groups of passengers - before we do anything with the data let's just check that the two sets are reasonably comparable.  By looking a a simple correlation matrix between the numeric collumns we should be able to see if the two groups are basically similar.

In [20]:
print ("Known :")
print (titanic_known[['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']].corr())
print ("\nUnknown :") 
print (titanic_unknown[['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']].corr())

Known :
          Pclass       Age     SibSp     Parch      Fare
Pclass  1.000000 -0.369226  0.083081  0.018443 -0.549500
Age    -0.369226  1.000000 -0.308247 -0.189119  0.096067
SibSp   0.083081 -0.308247  1.000000  0.414838  0.159651
Parch   0.018443 -0.189119  0.414838  1.000000  0.216225
Fare   -0.549500  0.096067  0.159651  0.216225  1.000000

Unknown :
          Pclass       Age     SibSp     Parch      Fare
Pclass  1.000000 -0.492143  0.001087  0.018721 -0.577147
Age    -0.492143  1.000000 -0.091587 -0.061249  0.337932
SibSp   0.001087 -0.091587  1.000000  0.306895  0.171539
Parch   0.018721 -0.061249  0.306895  1.000000  0.230046
Fare   -0.577147  0.337932  0.171539  0.230046  1.000000


This shows that our two groups are somewhat similar - but we have only looked at the numeric values.  This correlation ignores all the missing numeric values - and we know that a lot of ages are missing - a simple way to extend this analysis would be to encode gender and embarkation as integers and this would give us some additional information.

However we will be doing this anyway at a later stage so we can look at the revised correlation then.  The purpose of this exercise is just familiarity with the data.

